File: src/Astbox.Core/Extractor.cs
// Copyright 2026 Astenyx-Git
// SPDX-License-Identifier: MPL-2.0 OR AGPL-3.0-only
//! Extract files from an unlocked ASTBOX container to a local directory
//! (port of astbox/extract.py). Path safety: names were already validated at
//! metadata level; we still normalize and double-check containment.

namespace Astbox;

public static class Extractor
{
    public delegate void EntryProgress(Entry entry, ulong written, ulong size);
    public delegate void AllProgress(string stage, int done, int total);

    private static string SafeJoin(string outDir, IEnumerable<string> relParts)
    {
        string baseDir = Path.GetFullPath(outDir);
        string target = Path.GetFullPath(
            Path.Combine(new[] { outDir }.Concat(relParts).ToArray()));
        if (!target.Equals(baseDir, StringComparison.OrdinalIgnoreCase) &&
            !target.StartsWith(baseDir + Path.DirectorySeparatorChar,
                StringComparison.OrdinalIgnoreCase))
            throw new AstboxError(E.Write,
                "refusing to write outside output dir");
        return target;
    }

    /// <summary>Extract one file (or create one directory) below out_dir.
    /// Returns the absolute path written.</summary>
    public static string ExtractEntry(UnlockedContainer uc, Entry entry,
        string outDir, EntryProgress? progress = null)
    {
        var rel = Container.EntryPathParts(uc, entry);
        string target = SafeJoin(outDir, rel);
        if (entry.IsDir)
        {
            Directory.CreateDirectory(target);
            return target;
        }
        string? parent = Path.GetDirectoryName(target);
        try
        {
            if (!string.IsNullOrEmpty(parent)) Directory.CreateDirectory(parent);
            using var f = File.Create(target);
            ulong written = 0;
            foreach (var chunk in Container.IterFilePlaintext(uc, entry))
            {
                f.Write(chunk, 0, chunk.Length);
                written += (ulong)chunk.Length;
                progress?.Invoke(entry, written, entry.Size);
            }
            f.Flush(flushToDisk: true);
        }
        catch (Exception exc) when (exc is not AstboxError)
        {
            throw new AstboxError(E.Io,
                $"cannot write {target}: {exc.Message}");
        }
        return target;
    }

    /// <summary>Extract every file in the container below out_dir.</summary>
    public static List<(string LogicalPath, string AbsolutePath)> ExtractAll(
        UnlockedContainer uc, string outDir,
        AllProgress? progress = null, bool skipExisting = true)
    {
        var items = Container.WalkEntries(uc).Where(t => t.Entry.IsFile).ToList();
        int total = items.Count;
        var results = new List<(string, string)>(total);
        int i = 0;
        foreach (var (path, entry) in items)
        {
            i++;
            progress?.Invoke($"extracting {path} ({i}/{total})", i, total);
            string target = ExtractEntry(uc, entry, outDir);
            results.Add((path, target));
        }
        return results;
    }

    /// <summary>Extract a single file by logical path ('' → all).</summary>
    public static List<string> ExtractPath(UnlockedContainer uc,
        string logicalPath, string outDir)
    {
        if (string.IsNullOrEmpty(logicalPath))
            return ExtractAll(uc, outDir)
                .Select(r => r.AbsolutePath).ToList();
        foreach (var (p, e) in Container.WalkEntries(uc))
        {
            if (p != logicalPath) continue;
            if (e.IsDir)
                throw new AstboxError(E.Write,
                    $"'{logicalPath}' is a directory; extract its files " +
                    "individually");
            return new List<string> { ExtractEntry(uc, e, outDir) };
        }
        throw new AstboxError(E.Write, $"no such entry: '{logicalPath}'");
    }
}

File: rust/crates/astbox-core/src/extractor.rs
// Copyright 2026 Astenyx-Git
// SPDX-License-Identifier: MPL-2.0 OR AGPL-3.0-only
//! Extract files from an unlocked ASTBOX container to a local directory
//! (port of Astbox.Core/Extractor.cs). Path safety: names were already
//! validated at metadata level; we still normalize and double-check containment.

use crate::container::{cmp_ordinal, Container, Entry, UnlockedContainer};
use crate::errors::E;
use crate::Result;

pub type EntryProgress<'a> = dyn FnMut(&Entry, u64, u64) + 'a;
pub type AllProgress<'a> = dyn FnMut(&str, usize, usize) + 'a;

/// Lexically normalize a path (collapse `.` and `..`) without touching the
/// filesystem — the moral equivalent of C# Path.GetFullPath for our use.
fn normalize_lexical(p: &std::path::Path) -> std::path::PathBuf {
    use std::path::Component;
    let mut out = std::path::PathBuf::new();
    for comp in p.components() {
        match comp {
            Component::CurDir => {}
            Component::ParentDir => {
                out.pop();
            }
            other => out.push(other.as_os_str()),
        }
    }
    out
}

fn safe_join(out_dir: &str, rel_parts: &[String]) -> Result<std::path::PathBuf> {
    let base = normalize_lexical(std::path::Path::new(out_dir));
    let mut joined = std::path::PathBuf::from(out_dir);
    for part in rel_parts {
        joined.push(part);
    }
    let target = normalize_lexical(&joined);
    let base_s = base.to_string_lossy();
    let target_s = target.to_string_lossy();
    // Windows-style case-insensitive containment check (C# OrdinalIgnoreCase)
    let base_l = base_s.to_lowercase();
    let target_l = target_s.to_lowercase();
    let contained = target_l == base_l
        || target_l.starts_with(&format!("{}\\", base_l))
        || target_l.starts_with(&format!("{}/", base_l));
    if !contained {
        return Err(crate::err!(E::Write, "refusing to write outside output dir"));
    }
    Ok(target)
}

pub struct Extractor;

impl Extractor {
    /// Extract one file (or create one directory) below out_dir.
    /// Returns the absolute path written.
    pub fn extract_entry(
        uc: &UnlockedContainer,
        entry: &Entry,
        out_dir: &str,
        mut progress: Option<&mut EntryProgress>,
    ) -> Result<String> {
        let rel = Container::entry_path_parts(uc, entry);
        let target = safe_join(out_dir, &rel)?;
        if entry.is_dir() {
            std::fs::create_dir_all(&target)
                .map_err(|e| crate::err!(E::Io, "cannot create {}: {}", target.display(), e))?;
            return Ok(target.to_string_lossy().into_owned());
        }
        let parent = target.parent();
        if let Some(parent) = parent {
            std::fs::create_dir_all(parent).map_err(|e| {
                crate::err!(E::Io, "cannot create {}: {}", parent.display(), e)
            })?;
        }
        let mut written: u64 = 0;
        let chunks = Container::iter_file_plaintext(uc, entry)?;
        {
            use std::io::Write;
            let f = std::fs::File::create(&target).map_err(|e| {
                crate::err!(E::Io, "cannot write {}: {}", target.display(), e)
            })?;
            let mut w = std::io::BufWriter::new(f);
            for chunk in &chunks {
                w.write_all(chunk)
                    .map_err(|e| crate::err!(E::Io, "cannot write {}: {}", target.display(), e))?;
                written += chunk.len() as u64;
                if let Some(p) = progress.as_deref_mut() {
                    p(entry, written, entry.size);
                }
            }
            w.flush()
                .and_then(|_| w.into_inner().map_err(|e| e.into_error()))
                .and_then(|f| f.sync_all())
                .map_err(|e| crate::err!(E::Io, "cannot write {}: {}", target.display(), e))?;
        }
        Ok(target.to_string_lossy().into_owned())
    }

    /// Extract every file in the container below out_dir.
    /// `skip_existing` exists for signature parity with the C# original
    /// (which declares but never uses it).
    pub fn extract_all(
        uc: &UnlockedContainer,
        out_dir: &str,
        mut progress: Option<&mut AllProgress>,
        _skip_existing: bool,
    ) -> Result<Vec<(String, String)>> {
        let mut items: Vec<(String, Entry)> = Container::walk_entries(uc)
            .into_iter()
            .filter(|(_, e)| e.is_file())
            .collect();
        items.sort_by(|a, b| cmp_ordinal(&a.0, &b.0));
        let total = items.len();
        let mut results = Vec::with_capacity(total);
        for (i, (path, entry)) in items.iter().enumerate() {
            let i1 = i + 1;
            if let Some(p) = progress.as_deref_mut() {
                p(&format!("extracting {} ({}/{})", path, i1, total), i1, total);
            }
            let target = Self::extract_entry(uc, entry, out_dir, None)?;
            results.push((path.clone(), target));
        }
        Ok(results)
    }

    /// Extract a single file by logical path ('' → all).
    pub fn extract_path(
        uc: &UnlockedContainer,
        logical_path: &str,
        out_dir: &str,
    ) -> Result<Vec<String>> {
        if logical_path.is_empty() {
            return Ok(Self::extract_all(uc, out_dir, None, false)?
                .into_iter()
                .map(|(_, abs)| abs)
                .collect());
        }
        for (p, e) in Container::walk_entries(uc) {
            if p != logical_path {
                continue;
            }
            if e.is_dir() {
                return Err(crate::err!(
                    E::Write,
                    "'{}' is a directory; extract its files individually",
                    logical_path
                ));
            }
            return Ok(vec![Self::extract_entry(uc, &e, out_dir, None)?]);
        }
        Err(crate::err!(E::Write, "no such entry: '{}'", logical_path))
    }
}
